tachyon/zk/air/plonky3/base/two_adic_multiplicative_coset.h

```diff
@@ -0,0 +1,143 @@
+    return base::CreateVector(num_chunks, [log_chunks, this](size_t i) {
+      return TwoAdicMultiplicativeCoset(
+          domain_->log_size_of_group() - log_chunks,
+          domain_->offset() * domain_->group_gen().Pow(i));
+    });
+  }
+
+  template <typename ExtField>
+  air::plonky3::LagrangeSelectors<ExtField> GetSelectorsAtPoint(
+      const ExtField& point) const {
+    ExtField unshifted_point = point * domain_->offset_inv();
+    ExtField z_h = unshifted_point.ExpPowOfTwo(domain_->log_size_of_group()) -
+                   ExtField::One();
+    std::optional<ExtField> first_row =
+        z_h / (unshifted_point - ExtField::One());
+    CHECK(first_row);
+    ExtField transition = unshifted_point - ExtField(domain_->group_gen_inv());
+    std::optional<ExtField> last_row = z_h / transition;
+    CHECK(last_row);
+    ExtField inv_zeroifier = unwrap(z_h.Inverse());
+    return {std::move(*first_row), std::move(*last_row), std::move(transition),
+            std::move(inv_zeroifier)};
+  }
+
+  air::plonky3::LagrangeSelectors<std::vector<F>> GetSelectorsOnCoset(
+      const TwoAdicMultiplicativeCoset& coset) const {
+    F coset_shift = coset.domain()->offset();
+
+    CHECK_EQ(domain_->offset(), F::One());
+    CHECK_NE(coset_shift, F::One());
+    CHECK_GE(domain_->log_size_of_group(), coset.domain()->log_size_of_group());
+    size_t rate_bits =
+        coset.domain()->log_size_of_group() - domain_->log_size_of_group();
+    F s_pow_n = coset_shift.ExpPowOfTwo(domain_->log_size_of_group());
+
+    // Evals of Z_H(X) = Xⁿ - 1
+    std::vector<F> evals = base::Map(
+        domain_->GetRootsOfUnity(1 << rate_bits, domain_->group_gen()),
+        [&s_pow_n](F& x) { return s_pow_n * x - F::One(); });
+
+    std::vector<F> xs =
+        base::Map(domain_->GetRootsOfUnity(coset.domain()->size(),
+                                           coset.domain()->group_gen()),
+                  [&coset_shift](F& x) { return x * coset_shift; });
+    // FIRST_ROW
+    F coset_i = domain_->group_gen().Pow(0);
+    std::vector<F> inv_denoms =
+        base::Map(xs, [&coset_i](F& x) { return x - coset_i; });
+    CHECK(F::BatchInverseInPlace(inv_denoms));
+    std::vector<F> first_row =
+        base::CreateVector(inv_denoms.size(), [&evals, &inv_denoms](size_t i) {
+          return inv_denoms[i] * evals[i % evals.size()];
+        });
+    // LAST_ROW
+    coset_i = domain_->group_gen().Pow(domain_->size() - 1);
+
+    inv_denoms = base::Map(xs, [&coset_i](F& x) { return x - coset_i; });
+    CHECK(F::BatchInverseInPlace(inv_denoms));
+    std::vector<F> last_row =
+        base::CreateVector(inv_denoms.size(), [&evals, &inv_denoms](size_t i) {
+          return inv_denoms[i] * evals[i % evals.size()];
+        });
+    // TRANSITION
+    std::vector<F> transition =
+        base::Map(xs, [this](F& x) { return x - domain_->group_gen_inv(); });
+    // INV_ZEROIFIER
+    inv_denoms = evals;
```

Comment:
```suggestion
    inv_denoms = std::move(evals);
```

Reply:
This section of code is changed into base::Parallelize!